src/caliper/Annotation.cpp

```diff
@@ -136,35 +137,62 @@ Loop::end()
 // --- Annotation implementation object
 
 struct Annotation::Impl {
-    std::atomic<Attribute*> m_attr;
-    std::string             m_name;
-    int                     m_opt;
-    std::atomic<int>        m_refcount;
+    std::atomic<Attribute*>  m_attr;
+    std::string              m_name;
+    std::vector<cali_id_t>   m_metadata_keys;
+    std::vector<cali_variant_t>     m_metadata_values;
+    int                      m_opt;
+    std::atomic<int>         m_refcount;
+    Impl(const std::string& name, MetadataListType metadata, int opt)
+        : m_attr(nullptr), 
+          m_name(name), 
+          m_opt(opt),
+          m_refcount(1)
+        {
+            Caliper c;
+            if(!c.attribute_exists(m_name)){
+                for(auto kv : metadata){
+                    
+                    m_metadata_keys.push_back(cali_create_attribute(kv.first,kv.second.type(),0));
+                    m_metadata_values.push_back(kv.second.c_variant());
+                }
+            }
+        }
 
     Impl(const std::string& name, int opt)
         : m_attr(nullptr), 
           m_name(name), 
           m_opt(opt),
           m_refcount(1)
-        { }
+        {}
 
     ~Impl() {
         delete m_attr.load();
     }
-    
+    void establish_metadata(cali_attr_type type){
+        if(m_metadata_keys.empty()){
+            return;
+        }
+        cali_create_attribute_with_metadata(m_name.c_str(),type,m_opt,m_metadata_keys.size(),m_metadata_keys.data(),m_metadata_values.data());
+        m_metadata_keys.clear();
+    } 
     void begin(const Variant& data) {
+        cali_attr_type type = data.type();
+        establish_metadata(type);
```

Comment:
Why are you doing this in a new function now instead of in `get_attribute()`? You can just attach the metadata there in the `c.create_attribute(...)` call. Note that there are optional parameters to `Caliper::create_attribute()` for metadata.

Reply:
Great question, turns out that was a bad idea. Fixed